File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmQueryLetContext.java
package org.cqframework.cql.elm.requirements;

import org.hl7.elm.r1.*;

public class ElmQueryLetContext {
    public ElmQueryLetContext(VersionedIdentifier libraryIdentifier, LetClause letClause) {
        if (libraryIdentifier == null) {
            throw new IllegalArgumentException("libraryIdentifier is required");
        }
        if (letClause == null) {
            throw new IllegalArgumentException("letClause is required");
        }
        this.libraryIdentifier = libraryIdentifier;
        this.letClause = letClause;
    }

    private VersionedIdentifier libraryIdentifier;
    private LetClause letClause;

    public LetClause getLetClause() {
        return letClause;
    }

    public String getIdentifier() {
        return letClause.getIdentifier();
    }

    private ElmDataRequirement requirements;

    public ElmDataRequirement getRequirements() {
        return requirements;
    }

    public void setRequirements(ElmRequirement requirements) {
        if (requirements instanceof ElmDataRequirement) {
            this.requirements = (ElmDataRequirement) requirements;
        } else if (requirements instanceof ElmExpressionRequirement) {
            this.requirements = ElmDataRequirement.inferFrom((ElmExpressionRequirement) requirements);
        } else {
            // Should never land here, but defensively...
            this.requirements = new ElmDataRequirement(this.libraryIdentifier, new Retrieve());
        }
        this.requirements.setQuerySource(getLetClause());
    }

    public void reportProperty(ElmPropertyRequirement propertyRequirement) {
        requirements.reportProperty(propertyRequirement);
    }
}

File: elm-fhir/src/main/kotlin/org/cqframework/cql/elm/requirements/ElmQueryLetContext.kt
package org.cqframework.cql.elm.requirements

import org.hl7.elm.r1.LetClause
import org.hl7.elm.r1.Retrieve
import org.hl7.elm.r1.VersionedIdentifier

class ElmQueryLetContext(
    private val libraryIdentifier: VersionedIdentifier,
    @JvmField val letClause: LetClause,
) {

    val identifier: String?
        get() = letClause.identifier

    private var requirements: ElmDataRequirement? = null

    fun getRequirements(): ElmDataRequirement {
        return requirements!!
    }

    fun setRequirements(requirements: ElmRequirement?) {
        when (requirements) {
            is ElmDataRequirement -> {
                this.requirements = requirements
            }

            is ElmExpressionRequirement -> {
                this.requirements = ElmDataRequirement.inferFrom(requirements)
            }

            else -> {
                // Should never land here, but defensively...
                this.requirements = ElmDataRequirement(this.libraryIdentifier, Retrieve())
            }
        }
        this.requirements!!.querySource = this.letClause
    }

    fun reportProperty(propertyRequirement: ElmPropertyRequirement) {
        requirements!!.reportProperty(propertyRequirement)
    }
}
